packages/library-base/src/commonMain/kotlin/io/realm/internal/RealmObjectHelper.kt

```diff
@@ -168,25 +195,139 @@ internal object RealmObjectHelper {
+    ) {
+        if (target is DynamicRealmObject) {
+            assignDynamic(
+                target as DynamicMutableRealmObject,
+                source,
+                updatePolicy,
+                cache
+            )
+        } else {
+            assignTyped(target, source, updatePolicy, cache)
+        }
+    }
+
+    @Suppress("LongParameterList")
+    internal fun assignTyped(
+        target: BaseRealmObject,
+        source: BaseRealmObject,
+        updatePolicy: MutableRealm.UpdatePolicy,
+        cache: ObjectCache
+    ) {
+        val metadata: ClassMetadata = target.realmObjectReference!!.metadata
+        // TODO OPTIMIZE We could set all properties at once with on C-API call
+        for (property in metadata.properties) {
+            // Primary keys are set at construction time
+            if (property.isPrimaryKey) {
+                continue
+            }
+
+            val name = property.name
+            val accessor = property.acccessor ?: sdkError("Typed object should always have an accessor")
+            when (property.collectionType) {
+                CollectionType.RLM_COLLECTION_TYPE_NONE -> when (property.type) {
+                    PropertyType.RLM_PROPERTY_TYPE_OBJECT -> {
+                        setObject(
+                            target.realmObjectReference!!,
+                            name,
+                            accessor.get(source) as RealmObject?,
+                            updatePolicy,
+                            cache
+                        )
+                    }
+                    else ->
+                        accessor.set(target, accessor.get(source))
+                }
+                CollectionType.RLM_COLLECTION_TYPE_LIST -> {
+                    // We cannot use setList as that requires the type, so we need to retrieve the
+                    // existing list, wipe it and insert new elements
+                    @Suppress("UNCHECKED_CAST")
+                    (accessor.get(target) as ManagedRealmList<Any?>).run {
+                        clear()
+                        operator.insertAll(
+                            size,
+                            accessor.get(source) as RealmList<*>,
+                            updatePolicy,
+                            cache
+                        )
+                    }
+                }
+                else -> TODO("Collection type ${property.collectionType} is not supported")
             }
         }
     }
 
+    @Suppress("LongParameterList")
+    internal fun assignDynamic(
+        target: DynamicMutableRealmObject,
+        source: BaseRealmObject,
+        updatePolicy: MutableRealm.UpdatePolicy,
+        cache: ObjectCache
+    ) {
+        val properties: List<Pair<String, Any?>> = if (source is DynamicUnmanagedRealmObject) {
+            source.properties.toList()
+        } else if (source is DynamicRealmObject) {
+            TODO("Cannot import managed dynamic objects")
```

Comment:
We can actually not end up here. It was left over because I was imagining that I had to handle the case as part of importing dynamic embedded objects, but we should basically never end here. Added an explanation of why.